### ISSUE TITLE
🎛️ Migrate `initErrorDetection` to a Stimulus Controller `w-count` (tabs)

### ISSUE THREAD

**Issue by lb-** [wagtail/wagtail#10090]
> ℹ️ **Part of the [Stimulus  🎛️ RFC 78](https://github.com/wagtail/rfcs/pull/78)**

### Is your proposal related to a problem?

In the file `client/src/entrypoints/admin/page-editor.js` we have an ad-hoc JS implementation to check the count of some elements and update another element with the total number (icon/screen reader).

This is used to show the user visually that there are X errors within a tab in the page editor, and other editor interfaces. This code is now used across more than just page editor and is also a very common feature that we may want to use for other things such as the panels.

### Describe the solution you'd like

We should be able to replace the existing data attributes with Stimulus ones for a new controller `w-count` in the tabs file here `wagtail/admin/templates/wagtailadmin/shared/tabs/tab_nav_link.html`.

```html
<a id="tab-label-{{ tab_id }}" href="#tab-{{ tab_id }}" class="w-tabs__tab {{ classname }}" role="tab" aria-selected="false" tabindex="-1">
    {{ title }}
    <div class="w-tabs__errors" data-controller="w-count" data-w-count-active-class="!w-flex"  data-w-count-target-value="#tab-{{ tab_id }}">
        <span aria-hidden="true" data-w-count-target="total"></span>
        <span class="w-sr-only">(<span data-w-count-target="label"></span>)</span>
    </div>
</a>
```

This would be a new `CountController` that has defaults that assume we want to count errors but has the capacity to be more flexible in the long term.

The controller must have unit tests and preserve the existing tab count behaviour exactly as per current state.

Specification

* class: `active` (Stimulus class attribute) https://stimulus.hotwired.dev/reference/css-classes so that it can become visible when there is a count above 0
* value: `min` (number) - defaults to zero, will be used to determine when it is active
* value: `total` (number) - used as the 'state' of the current count and also helps for defaults on HTML load if needed
* value: `container` (string, optional, defaults to `body`) - a selector which could return multiple targets and is used to determine where to search for errors
* value: `target` (string, optional, defaults to `'.error-message,.help-critical,[aria-invalid="true"], .error, .w-field--error'`) used to determine the targets to count. Note: This is a collection of all similar 'error' selectors, check it is not too greedy.
* method: `count` - will find any `target` elements within the `container` element and returns the total AND updates the total value.
* target: `total` - inner text gets updated with the total when changed
* target: `label` - inner text gets updated with the translated label when changed
* method `totalValueChanged` - this will be a built in call-back  and should do the job of updating the translations
* method `getLabel` - just a wrapper for convenience on the labels.
* method `connect` - calls the `count` method but do this async (either in an immediate promise or set timeout), this small delay should account for picking up any DOM updates that are done by things like react or other JS adding errors to the DOM. We could put this behind a `delay` value of 0 if we wanted on the controller.

**Translations**

Remember to use the translations correctly. Long term we can add better support for non 'error' translations but that is probably getting a bit far out of scope.

However, let's share the same translation that has been set up more recently in MiniMapItem (for some reason it is different).

`client/src/components/Minimap/MinimapItem.tsx` uses `%(num)s` with brackets.

```js
  const errorsLabel = ngettext(
    '%(num)s error',
    '%(num)s errors',
    errorCount,
  ).replace('%(num)s', `${errorCount}`);
```

But, the older code, `client/src/entrypoints/admin/page-editor.js` uses `%(errorCount)s`, with brackets. We do need to support the brackets still though but lets put these in the HTML.

```js
    tabErrorsElement
      .find('[data-tabs-errors-statement]')
      .text(
        ngettext(
          '(%(errorCount)s error)',
          '(%(errorCount)s errors)',
          errorCount,
        ).replace('%(errorCount)s', errorCount),
      );
```

### Describe alternatives you've considered

* Leave as is, but makes it harder to support additional use cases and already there are duplicate implementations of this same behaviour.
* E.g. Similar behaviour in Panels here - `client/src/includes/panels.ts` - `const hasError = content.querySelector`
* E.g. in the minimap code that could be done with Stimulus. `client/src/components/Minimap/index.tsx`
```
  const errorCount = [].slice
    .call(panel.querySelectorAll('.error-message'))
    .filter((err) => err.closest('[data-panel]') === panel).length;
```

### Additional context

- This is part of a larger goal to migrate to Stimulus.
- A robust solution here may also be the starting point to solve for #8990 (in fact, once this ticket is done, that issue could probably be completed with just a few data attribute updates).
- Longer term, it would be good to support a more dynamic checking system.
- Not a good first issue but good for anyone semi-confident with Jest testing, JavaScript and who has taken the time to read the Stimulus docs in full.
- This is also part of a larger goal to avoid `wagtail/admin/templates/wagtailadmin/pages/_editor_js.html` being used everywhere and minimise global usage in JS See #2936

**Comment by lb-** [comment]
This is probably not a good first issue, but anyone who has read the Stimulus documentation end to end and takes a look at some other Controller implementations in `src/controllers` should be able make a solid go at this.

Note: Unit tests are a must - and it's actually best to start with unit tests to ensure your behaviour is working. One thing can be a bit tricky with DOM updates and settimeout usage, but see the tests for `ProgressController` and you should be able to get your head around it.

**Comment by salty-ivy** [comment]
would love to give it a try !

**Comment by lb-** [comment]
Thanks @salty-ivy  - there is no major urgency here, we are hopeful we can clean up the `page-editor.js` file in time for 5.0 but it depends on many other issues being done also.

I'm also hoping I will get the chance to get a few more 'smaller' type Stimulus adoption issues up in the coming weeks. It's a great way to learn the framework and also contribute something really helpful to the core code.

**Comment by salty-ivy** [comment]
@lb-  how would `totalValueChanged` update the translations in each tab differently, if `count` would directly count up the errors using `targetValue`, and we only get total number error elements only without any distinction of how many errors resides per tab.  shouldn't we count up the errors tab-wise just like its been counted in `initErrorDetection`.
or may be i am misunderstanding the explanation in the issue.

**Comment by lb-** [comment]
@salty-ivy great question - my proposal here is to actually have a separate 'instance' of the controller attached to each tab button and then that would only count the things in it's linked 'tab' content.

HTML structure

* Tabs container...
  * Tab A button - Main - `data-controller="w-count" data-w-count-container-value"#tab-a"` - pointing to the id of Tab A's content
  * Tab B button - Promote `data-controller="w-count" data-w-count-container-value"#tab-b"`
  * Tab A content - Main `div#tab-a`
  * Tab B content - Promote `div#tab-b`

So, by default `w-count` will just look through the entire `body` for elements, but in the tab's usage of the count controller, it will only be directed to look at it's associated tab's content.

The goal of this is to avoid the complexity of `initErrorDetection` needing to first count all the things and then also ensure they are grouped into 'tabs'. This controller approach will essentially only count one tab per instance, so no grouping needed.

This approach can then scale up/down to any kinds of counting containers (e.g. count the errors within just a panel like we do in other custom JS).

**Comment by salty-ivy** [comment]
ohh, i see its `#tab{{ tab_id }}` that's mapping instances of controller, got it.

**Comment by salty-ivy** [comment]
@lb-  the PR is ready, thanks a lot for the walk through into this, special thanks to @Lovelyfin00  for helping me out.
and apologies to both of you for the over all trouble, working through this new stuff i got prone to silly mistakes. 
although i understood the process pretty much and its going to be smoother and easier from here on out.

**Comment by lb-** [comment]
Awesome @salty-ivy - any feedback on the proposed contributor section for Stimulus would be appreciated. 

https://github.com/wagtail/wagtail/pull/10155

**Comment by salty-ivy** [comment]
Sure why not, let me give it a look.

**Comment by Lovelyfin00** [comment]
Great job @salty-ivy 👏👏👏
